Name : Jainam Rajput

SAP ID : 60009200022

Div/Batch : K/K1

In [33]:
import pandas as np
import numpy as np

**Question 1**

We make a Synthetic dataset using sklearn.datasets 

In [34]:
from sklearn.datasets import make_classification

In [35]:
x, y = make_classification(n_features=4, n_classes=2)

In [36]:
x,y

(array([[-8.72194580e-02,  7.78456594e-02, -7.02807081e-02,
         -1.46449007e+00],
        [ 2.83583814e-01, -8.73896891e-01, -1.59165051e+00,
          1.22234594e+00],
        [ 6.28437438e-02, -3.99775626e-02,  9.78797588e-02,
          1.14705972e+00],
        [-1.29178021e-01,  2.29574805e-01,  2.30978670e-01,
         -1.51747394e+00],
        [ 4.15345651e-01, -1.39968963e+00, -2.68229710e+00,
          1.10754569e+00],
        [ 3.50761557e-01, -1.18158660e+00, -2.26386936e+00,
          9.37942351e-01],
        [-1.22949608e-01,  5.80410161e-01,  1.28094642e+00,
          6.18992671e-01],
        [ 7.52820368e-02,  9.22380900e-02,  5.28108588e-01,
          2.17299342e+00],
        [-8.92075568e-02,  4.79658011e-01,  1.10102920e+00,
          7.82836072e-01],
        [-1.71736502e-01,  3.14843834e-01,  3.35323544e-01,
         -1.96248800e+00],
        [-2.08007816e-01,  6.94030900e-01,  1.32295558e+00,
         -5.94248686e-01],
        [-2.10938286e-02,  4.57438380e-01, 

Split the dataset

In [37]:
from sklearn.model_selection import train_test_split

In [38]:
xtrain, xtest, ytrain, ytest = train_test_split(x,y,test_size=0.20,random_state=41)

Logistic Regression Code From Scratch

In [39]:
class Logistic_Regression:
    def sigmoid(self,z):
        sig = 1/(1+np.exp(-z))
        return sig
    def initialize(self,X):
        weights = np.zeros((np.shape(X)[1]+1,1))
        X = np.c_[np.ones((np.shape(X)[0],1)),X]
        return weights,X
    def fit(self,X,y,alpha=0.001,iter=400):
        weights,X = self.initialize(X)
        def cost(theta):
            z = np.dot(X,theta)
            cost0 = y.T.dot(np.log(self.sigmoid(z)))
            cost1 = (1-y).T.dot(np.log(1-self.sigmoid(z)))
            cost = -((cost1 + cost0))/len(y)
            return cost
        cost_list = np.zeros(iter,)
        for i in range(iter):
            weights = weights - alpha*np.dot(X.T,self.sigmoid(np.dot(X,weights))-np.reshape(y,(len(y),1)))
            cost_list[i] = cost(weights)
        self.weights = weights
        return cost_list
    def predict(self,X):
        z = np.dot(self.initialize(X)[1],self.weights)
        lis = []
        for i in self.sigmoid(z):
            if i>0.5:
                lis.append(1)
            else:
                lis.append(0)
        return lis

In [40]:
def standardize(X_tr):
    for i in range(np.shape(X_tr)[1]):
        X_tr[:,i] = (X_tr[:,i] - np.mean(X_tr[:,i]))/np.std(X_tr[:,i])

In [41]:
standardize(xtrain)
standardize(xtest)

In [42]:
logreg = Logistic_Regression()
model = logreg.fit(xtrain,ytrain)

In [43]:
yhat_s = logreg.predict(xtest) 

In [44]:
yhat_s

[0, 0, 1, 1, 1, 1, 1, 0, 1, 0, 1, 0, 0, 1, 0, 1, 0, 0, 0, 1]

In [45]:
def f1Score(y,yhat):
  tp,tn,fp,fn = 0,0,0,0
  for i in range(len(y)):
      if y[i] == 1 and yhat[i] == 1:
          tp += 1
      elif y[i] == 1 and yhat[i] == 0:
          fn += 1
      elif y[i] == 0 and yhat[i] == 1:
          fp += 1
      elif y[i] == 0 and yhat[i] == 0:
          tn += 1
  precision = tp/(tp+fp)
  recall = tp/(tp+fn)
  f1_score = 2*precision*recall/(precision+recall)
  return f1_score

In [46]:
print(f1Score(ytest,yhat_s))

0.9523809523809523


Here we get F1 Score = 0.952 from our scratch model

In [47]:
from sklearn.metrics import f1_score, confusion_matrix, accuracy_score 

In [48]:
print(f1_score(ytest,yhat_s))

0.9523809523809523


In [49]:
cm = confusion_matrix(ytest,yhat_s)
cm

array([[ 9,  0],
       [ 1, 10]])

Checked For the correctness of F1 score function

Now we Train the model using built in library

In [50]:
from sklearn.linear_model import LogisticRegression

In [51]:
lr = LogisticRegression()
lr.fit(xtrain,ytrain)

LogisticRegression()

In [52]:
yhat = lr.predict(xtest)

In [53]:
print(f"The F1 score is {f1_score(ytest,yhat)}")

The F1 score is 0.9


We get the same f1 score from our model trained by built in library as trained by Logistic Regression Function from scratch

2nd Question

In [54]:
from sklearn.datasets import load_iris

In [55]:
iris = load_iris()
x = iris.data
y = iris.target

In [56]:
x,y

(array([[5.1, 3.5, 1.4, 0.2],
        [4.9, 3. , 1.4, 0.2],
        [4.7, 3.2, 1.3, 0.2],
        [4.6, 3.1, 1.5, 0.2],
        [5. , 3.6, 1.4, 0.2],
        [5.4, 3.9, 1.7, 0.4],
        [4.6, 3.4, 1.4, 0.3],
        [5. , 3.4, 1.5, 0.2],
        [4.4, 2.9, 1.4, 0.2],
        [4.9, 3.1, 1.5, 0.1],
        [5.4, 3.7, 1.5, 0.2],
        [4.8, 3.4, 1.6, 0.2],
        [4.8, 3. , 1.4, 0.1],
        [4.3, 3. , 1.1, 0.1],
        [5.8, 4. , 1.2, 0.2],
        [5.7, 4.4, 1.5, 0.4],
        [5.4, 3.9, 1.3, 0.4],
        [5.1, 3.5, 1.4, 0.3],
        [5.7, 3.8, 1.7, 0.3],
        [5.1, 3.8, 1.5, 0.3],
        [5.4, 3.4, 1.7, 0.2],
        [5.1, 3.7, 1.5, 0.4],
        [4.6, 3.6, 1. , 0.2],
        [5.1, 3.3, 1.7, 0.5],
        [4.8, 3.4, 1.9, 0.2],
        [5. , 3. , 1.6, 0.2],
        [5. , 3.4, 1.6, 0.4],
        [5.2, 3.5, 1.5, 0.2],
        [5.2, 3.4, 1.4, 0.2],
        [4.7, 3.2, 1.6, 0.2],
        [4.8, 3.1, 1.6, 0.2],
        [5.4, 3.4, 1.5, 0.4],
        [5.2, 4.1, 1.5, 0.1],
        [5

Here in the dataset column to be poredicted i.e. y has 3 unique values. Hence we prefer a Multinomial Model

In [57]:
from sklearn.model_selection import train_test_split

In [58]:
x_train, x_test, y_train, y_test = train_test_split(x,y,test_size=0.20,random_state=41)

In [59]:
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()

In [60]:
x_train = sc.fit_transform(x_train)
x_test = sc.fit_transform(x_test)

In [61]:
x_train,y_train

(array([[-0.18688501, -1.09883199, -0.14183598, -0.25429341],
        [ 0.77252466, -0.6472572 ,  0.47403077,  0.39496637],
        [-1.14629467, -1.32461939,  0.41804288,  0.65467028],
        [-0.42673742, -1.09883199,  0.362055  ,  0.0054105 ],
        [ 1.01237707, -0.19568241,  0.80995808,  1.43378202],
        [-0.0669588 ,  2.06219154, -1.42955736, -1.29310906],
        [ 1.25222949,  0.03010499,  0.7539702 ,  1.43378202],
        [ 0.65259845,  0.25589238,  0.41804288,  0.39496637],
        [ 1.25222949,  0.03010499,  0.92193386,  1.1740781 ],
        [ 0.05296741,  0.25589238,  0.58600654,  0.78452224],
        [ 0.41274603, -0.4214698 ,  0.30606711,  0.13526245],
        [-0.18688501, -0.6472572 ,  0.41804288,  0.13526245],
        [-0.78651605,  2.28797894, -1.2615937 , -1.42296102],
        [ 0.77252466, -0.19568241,  1.1458854 ,  1.30393006],
        [ 0.53267224,  0.70746717,  1.03390963,  1.56363397],
        [-0.18688501,  1.61061675, -1.14961793, -1.16325711],
        

In [62]:
from sklearn.linear_model import LogisticRegression

In [63]:
lr = LogisticRegression(multi_class='multinomial',solver='lbfgs')
lr.fit(x_train,y_train)

LogisticRegression(multi_class='multinomial')

In [64]:
yhat = lr.predict(x_test)
yhat

array([1, 2, 2, 1, 2, 1, 2, 1, 2, 2, 2, 0, 0, 1, 0, 1, 0, 2, 0, 0, 1, 2,
       0, 0, 1, 1, 1, 1, 0, 1])

In [65]:
from sklearn.metrics import accuracy_score,confusion_matrix,f1_score

In [66]:
print(f"The Accuracy of the model is {accuracy_score(y_test,yhat)}")
cm = confusion_matrix(y_test,yhat)
print(cm)
print(f"The F1 score of the model is {f1_score(y_test,yhat,average='macro')}")

The Accuracy of the model is 0.9
[[ 9  0  0]
 [ 0 10  1]
 [ 0  2  8]]
The F1 score of the model is 0.9038901601830663


Here from the model we get a Accuracy of 90% and F1 score of 0.9038 on test data

3rd Question

In [67]:
import pandas as pd
import numpy as np

In [69]:
df = pd.read_csv('/content/train.csv')
df1 = pd.read_csv('/content/test.csv')

In [70]:
df.head()

Unnamed: 0      id  Gender      Customer Type  Age   Type of Travel  \
0           0   70172    Male     Loyal Customer   13  Personal Travel   
1           1    5047    Male  disloyal Customer   25  Business travel   
2           2  110028  Female     Loyal Customer   26  Business travel   
3           3   24026  Female     Loyal Customer   25  Business travel   
4           4  119299    Male     Loyal Customer   61  Business travel   

      Class  Flight Distance  Inflight wifi service  \
0  Eco Plus            460.0                    3.0   
1  Business            235.0                    3.0   
2  Business           1142.0                    2.0   
3  Business            562.0                    2.0   
4  Business            214.0                    3.0   

   Departure/Arrival time convenient  ...  Inflight entertainment  \
0                                4.0  ...                     5.0   
1                                2.0  ...                     1.0   
2                                2.0  ...                     5.0   
3                                5.0  ...                     2.0   
4                                3.0  ...                     3.0   

   On-board service  Leg room service  Baggage handling  Checkin service  \
0               4.0               3.0               4.0              4.0   
1               1.0               5.0               3.0              1.0   
2               4.0               3.0               4.0              4.0   
3               2.0               5.0               3.0              1.0   
4               3.0               4.0               4.0              3.0   

   Inflight service  Cleanliness  Departure Delay in Minutes  \
0               5.0          5.0                        25.0   
1               4.0          1.0                         1.0   
2               4.0          5.0                         0.0   
3               4.0          2.0                        11.0   
4               3.0          3.0                         0.0   

   Arrival Delay in Minutes             satisfaction  
0                      18.0  neutral or dissatisfied  
1                       6.0  neutral or dissatisfied  
2                       0.0                satisfied  
3                       9.0  neutral or dissatisfied  
4                       0.0                satisfied  

[5 rows x 25 columns]

In [71]:
df1.head()

Unnamed: 0     id  Gender      Customer Type  Age   Type of Travel  \
0           0  19556  Female     Loyal Customer   52  Business travel   
1           1  90035  Female     Loyal Customer   36  Business travel   
2           2  12360    Male  disloyal Customer   20  Business travel   
3           3  77959    Male     Loyal Customer   44  Business travel   
4           4  36875  Female     Loyal Customer   49  Business travel   

      Class  Flight Distance  Inflight wifi service  \
0       Eco              160                      5   
1  Business             2863                      1   
2       Eco              192                      2   
3  Business             3377                      0   
4       Eco             1182                      2   

   Departure/Arrival time convenient  ...  Inflight entertainment  \
0                                  4  ...                       5   
1                                  1  ...                       4   
2                                  0  ...                       2   
3                                  0  ...                       1   
4                                  3  ...                       2   

   On-board service  Leg room service  Baggage handling  Checkin service  \
0                 5                 5                 5                2   
1                 4                 4                 4                3   
2                 4                 1                 3                2   
3                 1                 1                 1                3   
4                 2                 2                 2                4   

   Inflight service  Cleanliness  Departure Delay in Minutes  \
0                 5            5                          50   
1                 4            5                           0   
2                 2            2                           0   
3                 1            4                           0   
4                 2            4                           0   

   Arrival Delay in Minutes             satisfaction  
0                      44.0                satisfied  
1                       0.0                satisfied  
2                       0.0  neutral or dissatisfied  
3                       6.0                satisfied  
4                      20.0                satisfied  

[5 rows x 25 columns]

In [72]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 44765 entries, 0 to 44764
Data columns (total 25 columns):
 #   Column                             Non-Null Count  Dtype  
---  ------                             --------------  -----  
 0   Unnamed: 0                         44765 non-null  int64  
 1   id                                 44765 non-null  int64  
 2   Gender                             44765 non-null  object 
 3   Customer Type                      44765 non-null  object 
 4   Age                                44765 non-null  int64  
 5   Type of Travel                     44765 non-null  object 
 6   Class                              44764 non-null  object 
 7   Flight Distance                    44764 non-null  float64
 8   Inflight wifi service              44764 non-null  float64
 9   Departure/Arrival time convenient  44764 non-null  float64
 10  Ease of Online booking             44764 non-null  float64
 11  Gate location                      44764 non-null  flo

Performing Data Cleaning on both the train and test data

In [73]:
df.isna().sum()

Unnamed: 0                             0
id                                     0
Gender                                 0
Customer Type                          0
Age                                    0
Type of Travel                         0
Class                                  1
Flight Distance                        1
Inflight wifi service                  1
Departure/Arrival time convenient      1
Ease of Online booking                 1
Gate location                          1
Food and drink                         1
Online boarding                        1
Seat comfort                           1
Inflight entertainment                 1
On-board service                       1
Leg room service                       1
Baggage handling                       1
Checkin service                        1
Inflight service                       1
Cleanliness                            1
Departure Delay in Minutes             1
Arrival Delay in Minutes             137
satisfaction    

There are 137 null values in arrival delay in minutes and 1 null value in some other columns

In [74]:
df1.isna().sum()

Unnamed: 0                            0
id                                    0
Gender                                0
Customer Type                         0
Age                                   0
Type of Travel                        0
Class                                 0
Flight Distance                       0
Inflight wifi service                 0
Departure/Arrival time convenient     0
Ease of Online booking                0
Gate location                         0
Food and drink                        0
Online boarding                       0
Seat comfort                          0
Inflight entertainment                0
On-board service                      0
Leg room service                      0
Baggage handling                      0
Checkin service                       0
Inflight service                      0
Cleanliness                           0
Departure Delay in Minutes            0
Arrival Delay in Minutes             83
satisfaction                          0


There are 83 null values in Test dataset

In [75]:
df['Arrival Delay in Minutes'].fillna(df['Arrival Delay in Minutes'].mean(),inplace=True)
df1['Arrival Delay in Minutes'].fillna(df1['Arrival Delay in Minutes'].mean(),inplace=True)

In [76]:
df.head()

Unnamed: 0      id  Gender      Customer Type  Age   Type of Travel  \
0           0   70172    Male     Loyal Customer   13  Personal Travel   
1           1    5047    Male  disloyal Customer   25  Business travel   
2           2  110028  Female     Loyal Customer   26  Business travel   
3           3   24026  Female     Loyal Customer   25  Business travel   
4           4  119299    Male     Loyal Customer   61  Business travel   

      Class  Flight Distance  Inflight wifi service  \
0  Eco Plus            460.0                    3.0   
1  Business            235.0                    3.0   
2  Business           1142.0                    2.0   
3  Business            562.0                    2.0   
4  Business            214.0                    3.0   

   Departure/Arrival time convenient  ...  Inflight entertainment  \
0                                4.0  ...                     5.0   
1                                2.0  ...                     1.0   
2                                2.0  ...                     5.0   
3                                5.0  ...                     2.0   
4                                3.0  ...                     3.0   

   On-board service  Leg room service  Baggage handling  Checkin service  \
0               4.0               3.0               4.0              4.0   
1               1.0               5.0               3.0              1.0   
2               4.0               3.0               4.0              4.0   
3               2.0               5.0               3.0              1.0   
4               3.0               4.0               4.0              3.0   

   Inflight service  Cleanliness  Departure Delay in Minutes  \
0               5.0          5.0                        25.0   
1               4.0          1.0                         1.0   
2               4.0          5.0                         0.0   
3               4.0          2.0                        11.0   
4               3.0          3.0                         0.0   

   Arrival Delay in Minutes             satisfaction  
0                      18.0  neutral or dissatisfied  
1                       6.0  neutral or dissatisfied  
2                       0.0                satisfied  
3                       9.0  neutral or dissatisfied  
4                       0.0                satisfied  

[5 rows x 25 columns]

In [77]:
df.dropna(inplace=True)

In [78]:
df.isna().sum()

Unnamed: 0                           0
id                                   0
Gender                               0
Customer Type                        0
Age                                  0
Type of Travel                       0
Class                                0
Flight Distance                      0
Inflight wifi service                0
Departure/Arrival time convenient    0
Ease of Online booking               0
Gate location                        0
Food and drink                       0
Online boarding                      0
Seat comfort                         0
Inflight entertainment               0
On-board service                     0
Leg room service                     0
Baggage handling                     0
Checkin service                      0
Inflight service                     0
Cleanliness                          0
Departure Delay in Minutes           0
Arrival Delay in Minutes             0
satisfaction                         0
dtype: int64

In [79]:
df1.isna().sum()

Unnamed: 0                           0
id                                   0
Gender                               0
Customer Type                        0
Age                                  0
Type of Travel                       0
Class                                0
Flight Distance                      0
Inflight wifi service                0
Departure/Arrival time convenient    0
Ease of Online booking               0
Gate location                        0
Food and drink                       0
Online boarding                      0
Seat comfort                         0
Inflight entertainment               0
On-board service                     0
Leg room service                     0
Baggage handling                     0
Checkin service                      0
Inflight service                     0
Cleanliness                          0
Departure Delay in Minutes           0
Arrival Delay in Minutes             0
satisfaction                         0
dtype: int64

In [80]:
df.drop(['Unnamed: 0','id','Gender','Arrival Delay in Minutes'],axis=1,inplace=True)
df1.drop(['Unnamed: 0','id','Gender','Arrival Delay in Minutes'],axis=1,inplace=True)

Dropped the unnnecessary columns from both the datasets

In [81]:
df.head()

Customer Type  Age   Type of Travel     Class  Flight Distance  \
0     Loyal Customer   13  Personal Travel  Eco Plus            460.0   
1  disloyal Customer   25  Business travel  Business            235.0   
2     Loyal Customer   26  Business travel  Business           1142.0   
3     Loyal Customer   25  Business travel  Business            562.0   
4     Loyal Customer   61  Business travel  Business            214.0   

   Inflight wifi service  Departure/Arrival time convenient  \
0                    3.0                                4.0   
1                    3.0                                2.0   
2                    2.0                                2.0   
3                    2.0                                5.0   
4                    3.0                                3.0   

   Ease of Online booking  Gate location  Food and drink  ...  Seat comfort  \
0                     3.0            1.0             5.0  ...           5.0   
1                     3.0            3.0             1.0  ...           1.0   
2                     2.0            2.0             5.0  ...           5.0   
3                     5.0            5.0             2.0  ...           2.0   
4                     3.0            3.0             4.0  ...           5.0   

   Inflight entertainment  On-board service  Leg room service  \
0                     5.0               4.0               3.0   
1                     1.0               1.0               5.0   
2                     5.0               4.0               3.0   
3                     2.0               2.0               5.0   
4                     3.0               3.0               4.0   

   Baggage handling  Checkin service  Inflight service  Cleanliness  \
0               4.0              4.0               5.0          5.0   
1               3.0              1.0               4.0          1.0   
2               4.0              4.0               4.0          5.0   
3               3.0              1.0               4.0          2.0   
4               4.0              3.0               3.0          3.0   

   Departure Delay in Minutes             satisfaction  
0                        25.0  neutral or dissatisfied  
1                         1.0  neutral or dissatisfied  
2                         0.0                satisfied  
3                        11.0  neutral or dissatisfied  
4                         0.0                satisfied  

[5 rows x 21 columns]

In [82]:
df1.head()

Customer Type  Age   Type of Travel     Class  Flight Distance  \
0     Loyal Customer   52  Business travel       Eco              160   
1     Loyal Customer   36  Business travel  Business             2863   
2  disloyal Customer   20  Business travel       Eco              192   
3     Loyal Customer   44  Business travel  Business             3377   
4     Loyal Customer   49  Business travel       Eco             1182   

   Inflight wifi service  Departure/Arrival time convenient  \
0                      5                                  4   
1                      1                                  1   
2                      2                                  0   
3                      0                                  0   
4                      2                                  3   

   Ease of Online booking  Gate location  Food and drink  ...  Seat comfort  \
0                       3              4               3  ...             3   
1                       3              1               5  ...             5   
2                       2              4               2  ...             2   
3                       0              2               3  ...             4   
4                       4              3               4  ...             2   

   Inflight entertainment  On-board service  Leg room service  \
0                       5                 5                 5   
1                       4                 4                 4   
2                       2                 4                 1   
3                       1                 1                 1   
4                       2                 2                 2   

   Baggage handling  Checkin service  Inflight service  Cleanliness  \
0                 5                2                 5            5   
1                 4                3                 4            5   
2                 3                2                 2            2   
3                 1                3                 1            4   
4                 2                4                 2            4   

   Departure Delay in Minutes             satisfaction  
0                          50                satisfied  
1                           0                satisfied  
2                           0  neutral or dissatisfied  
3                           0                satisfied  
4                           0                satisfied  

[5 rows x 21 columns]

In [83]:
from sklearn.preprocessing import LabelEncoder

In [84]:
le = LabelEncoder()
df['Customer Type'] = le.fit_transform(df['Customer Type'])

In [85]:
df['Class'] = le.fit_transform(df['Class'])
df['satisfaction'] = le.fit_transform(df['satisfaction'])
df['Type of Travel'] = le.fit_transform(df['Type of Travel'])

In [86]:
df.head()

Customer Type  Age  Type of Travel  Class  Flight Distance  \
0              0   13               1      2            460.0   
1              1   25               0      0            235.0   
2              0   26               0      0           1142.0   
3              0   25               0      0            562.0   
4              0   61               0      0            214.0   

   Inflight wifi service  Departure/Arrival time convenient  \
0                    3.0                                4.0   
1                    3.0                                2.0   
2                    2.0                                2.0   
3                    2.0                                5.0   
4                    3.0                                3.0   

   Ease of Online booking  Gate location  Food and drink  ...  Seat comfort  \
0                     3.0            1.0             5.0  ...           5.0   
1                     3.0            3.0             1.0  ...           1.0   
2                     2.0            2.0             5.0  ...           5.0   
3                     5.0            5.0             2.0  ...           2.0   
4                     3.0            3.0             4.0  ...           5.0   

   Inflight entertainment  On-board service  Leg room service  \
0                     5.0               4.0               3.0   
1                     1.0               1.0               5.0   
2                     5.0               4.0               3.0   
3                     2.0               2.0               5.0   
4                     3.0               3.0               4.0   

   Baggage handling  Checkin service  Inflight service  Cleanliness  \
0               4.0              4.0               5.0          5.0   
1               3.0              1.0               4.0          1.0   
2               4.0              4.0               4.0          5.0   
3               3.0              1.0               4.0          2.0   
4               4.0              3.0               3.0          3.0   

   Departure Delay in Minutes  satisfaction  
0                        25.0             0  
1                         1.0             0  
2                         0.0             1  
3                        11.0             0  
4                         0.0             1  

[5 rows x 21 columns]

In [87]:
df1['Customer Type'] = le.fit_transform(df1['Customer Type'])
df1['Class'] = le.fit_transform(df1['Class'])
df1['satisfaction'] = le.fit_transform(df1['satisfaction'])
df1['Type of Travel'] = le.fit_transform(df1['Type of Travel'])

In [88]:
df1.head()

Customer Type  Age  Type of Travel  Class  Flight Distance  \
0              0   52               0      1              160   
1              0   36               0      0             2863   
2              1   20               0      1              192   
3              0   44               0      0             3377   
4              0   49               0      1             1182   

   Inflight wifi service  Departure/Arrival time convenient  \
0                      5                                  4   
1                      1                                  1   
2                      2                                  0   
3                      0                                  0   
4                      2                                  3   

   Ease of Online booking  Gate location  Food and drink  ...  Seat comfort  \
0                       3              4               3  ...             3   
1                       3              1               5  ...             5   
2                       2              4               2  ...             2   
3                       0              2               3  ...             4   
4                       4              3               4  ...             2   

   Inflight entertainment  On-board service  Leg room service  \
0                       5                 5                 5   
1                       4                 4                 4   
2                       2                 4                 1   
3                       1                 1                 1   
4                       2                 2                 2   

   Baggage handling  Checkin service  Inflight service  Cleanliness  \
0                 5                2                 5            5   
1                 4                3                 4            5   
2                 3                2                 2            2   
3                 1                3                 1            4   
4                 2                4                 2            4   

   Departure Delay in Minutes  satisfaction  
0                          50             1  
1                           0             1  
2                           0             0  
3                           0             1  
4                           0             1  

[5 rows x 21 columns]

In [89]:
x = df.iloc[:,:-1].values
y = df.iloc[:,-1].values

In [90]:
x

array([[ 0., 13.,  1., ...,  5.,  5., 25.],
       [ 1., 25.,  0., ...,  4.,  1.,  1.],
       [ 0., 26.,  0., ...,  4.,  5.,  0.],
       ...,
       [ 0., 29.,  0., ...,  3.,  2., 20.],
       [ 0., 29.,  0., ...,  4.,  5.,  0.],
       [ 0., 28.,  0., ...,  1.,  4.,  0.]])

In [91]:
y

array([0, 0, 1, ..., 0, 1, 1])

In [92]:
x1 = df1.iloc[:,:-1].values
y1 = df1.iloc[:,-1].values

In [93]:
from sklearn.model_selection import train_test_split

In [108]:
x_train, x_test, y_train, y_test = train_test_split(x,y,test_size=0.30)

In [109]:
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
x_train = sc.fit_transform(x_train)
x_test = sc.fit_transform(x_test)

In [110]:
x = sc.fit_transform(x)

In [111]:
from sklearn.linear_model import LogisticRegression

Without Regularization

In [112]:
lr = LogisticRegression()
lr.fit(x_train,y_train)

LogisticRegression()

In [113]:
yhat_train_data = lr.predict(x_test)
yhat_test_data = lr.predict(x)

In [114]:
from sklearn.metrics import accuracy_score,f1_score,confusion_matrix

In [115]:
print(f"The Accuracy of the model on train dataset is {accuracy_score(y_test,yhat_train_data)}")
print(f"The Accuracy of the model on test dataset is {accuracy_score(y,yhat_test_data)}")

The Accuracy of the model on train dataset is 0.8719285182427401
The Accuracy of the model on test dataset is 0.873246358681083


In [116]:
print(f"The F1 score of the model on train dataset is {f1_score(y_test,yhat_train_data)}")
print(f"The F1 score of the model on test dataset is {f1_score(y,yhat_test_data)}")

The F1 score of the model on train dataset is 0.8511853261810001
The F1 score of the model on test dataset is 0.8521009279532896


In [117]:
cm = confusion_matrix(y_test,yhat_train_data)
print(f"Train Data\n{cm}")
cm1 = confusion_matrix(y,yhat_test_data)
print(f"Test Data\n{cm1}")

Train Data
[[6791  730]
 [ 990 4919]]
Test Data
[[22745  2421]
 [ 3253 16345]]


**With Regularization**

L1 Regularization

In [122]:
lr_l1 = LogisticRegression(penalty='l1',solver='liblinear')
lr_l1.fit(x_train,y_train)
yhat_train_data1 = lr_l1.predict(x_test)
yhat_test_data1 = lr_l1.predict(x)

In [124]:
print(f"The Accuracy of the model on train dataset is {f1_score(y_test,yhat_train_data1)}")
print(f"The Accuracy of the model on test dataset is {f1_score(y,yhat_test_data1)}")

The Accuracy of the model on train dataset is 0.8511853261810001
The Accuracy of the model on test dataset is 0.8520043788771308


L2 Regularization

In [121]:
lr_l2 = LogisticRegression(penalty='l2')
lr_l2.fit(x_train,y_train)
yhat_train_data2 = lr_l2.predict(x_test)
yhat_test_data2 = lr_l2.predict(x)

In [125]:
print(f"The Accuracy of the model on train dataset is {f1_score(y_test,yhat_train_data2)}")
print(f"The Accuracy of the model on test dataset is {f1_score(y,yhat_test_data2)}")

The Accuracy of the model on train dataset is 0.8511853261810001
The Accuracy of the model on test dataset is 0.8521009279532896


L1+L2 Regularization

In [128]:
lr_l3 = LogisticRegression(penalty='elasticnet',solver='saga',max_iter=116,l1_ratio=1)
lr_l3.fit(x_train,y_train)

LogisticRegression(l1_ratio=1, max_iter=116, penalty='elasticnet',
                   solver='saga')

In [129]:
yhat_train_data3 = lr_l3.predict(x_test)
yhat_test_data3 = lr_l3.predict(x)

In [130]:
print(f"The Accuracy of the model on train dataset is {f1_score(y_test,yhat_train_data3)}")
print(f"The Accuracy of the model on test dataset is {f1_score(y,yhat_test_data3)}")

The Accuracy of the model on train dataset is 0.8511853261810001
The Accuracy of the model on test dataset is 0.8520043788771308
